core/src/block.rs

```diff
@@ -695,23 +694,34 @@ mod valid {
             .map_err(|_| InvalidGenesisError::InvalidSignature)?;
 
         let transactions = block.payload().transactions.as_slice();
-        let [transaction] = transactions else {
-            return Err(InvalidGenesisError::NotSingleTransaction);
-        };
-        if transaction.value.authority() != genesis_account {
-            return Err(InvalidGenesisError::UnexpectedAuthority);
-        }
-
-        if transaction.error.is_some() {
-            return Err(InvalidGenesisError::ContainsErrors);
+        for transaction in transactions {
+            if transaction.value.authority() != genesis_account {
+                return Err(InvalidGenesisError::UnexpectedAuthority);
+            }
+            if transaction.error.is_some() {
+                return Err(InvalidGenesisError::ContainsErrors);
+            }
+            let Executable::Instructions(_) = transaction.value.instructions() else {
+                return Err(InvalidGenesisError::InvalidInstructions);
+            };
```

Comment:
I think you could have put many of these checks into `SignedBlock::deserialize` and reject eagerly